### PR TITLE
[MCC-114510] Added AttributeConstraint.

### PR DESCRIPTION
**Pull request by kenyamat**
@mdsol/hypermedia-team Added AttributeConstraint. Please review this.

### CODE REVIEW

File: tests/Crichton.Representors.Tests/Serializers/HaleSerializerTests.cs [hunk 1]
```diff
@@ -169,6 +169,23 @@ public void Serialize_AddsJsonTypeForEachTransition()
...
+                    var attr = result["_links"][transition.Rel]["data"][attribute.Key];
+                    var options = attr["options"].Values<string>().ToList();
+
+                    Assert.AreEqual(attribute.Value.Constraint.Options.Count, options.Count);
+                    for (int i = 0; i < attribute.Value.Constraint.Options.Count; i++)
+                    {
+                        Assert.AreEqual(attribute.Value.Constraint.Options[i], options[i]);
+                    }
+
+                    Assert.AreEqual(attribute.Value.Constraint.IsIn, attr["in"].Value<bool?>());
```
Comment (edandersen) [inline tests/Crichton.Representors.Tests/Serializers/HaleSerializerTests.cs:182]: @kenyamat It looks like you've gone back and just added to existing tests. Now this test, which originally only tested the JsonType property, is now testing too much, including the entire Constraint feature. The problem with this is that if any of the constraint assignment operations fails, this test will fail but we won't be able to tell which feature doesn't work. There should be a test for each constraint property. 

Please create a test for each Constraint property.

File: tests/Crichton.Representors.Tests/Serializers/HaleSerializerTests.cs [hunk 2]
```diff
@@ -775,5 +792,85 @@ public void DeserializeToNewBuilder_DeserializesNestedDataObjectsWithEitherScope
...
+            }}";
+
+            json = String.Format(json, option0, option1, isIn, min, minLength, max, maxLength, pattern, isMulti, isRequired);
+
+            var builder = sut.DeserializeToNewBuilder(json, builderFactoryMethod);
+
+            var transition = (CrichtonTransition)(builder.GetArgumentsForCallsMadeOn(b => b.AddTransition(null))[0].First());
+            var constraints = transition.Attributes.First().Value.Constraint;
+
+            Assert.AreEqual(2, constraints.Options.Count);
```
Comment (edandersen) [inline tests/Crichton.Representors.Tests/Serializers/HaleSerializerTests.cs:1097]: Same applies here - there is value in this test but it only confirms the system works when every single Constraint property is populated. Please split this up.

**Comment by edandersen** [comment]
Please address comments then we can merge :thumbsup: Thank you!

**Comment by kenyamat** [comment]
@edandersen Thank you for your feedback. I separated unit tests. Please review this again.

**Comment by edandersen** [comment]
Great stuff. Merging.

**Pull request merged**